### PR TITLE
Added new file full of session variables, plus new types

### PR DESCRIPTION
**Pull request by Hydrocharged**
I'm sure there are mistakes in here. This was like 85% manually done, and I had lots of mental TODOs while working through it, so I very well may have forgotten to edit a variable, add a note, or change the type. Either way, the majority of variables are here. I did exclude some as I couldn't find any `GLOBAL/SESSION` or `Dynamic` information for them, but this is like 98% of the variables.

Now to hook them up and replace our current variable handling.

### CODE REVIEW

File: sql/system_variables.go [hunk 1]
```diff
@@ -0,0 +1,2459 @@
...
+	"time_zone": {
+		Name: "time_zone",
+		Scope: SystemVariableScope_Both,
+		Dynamic: true,
+		SetVarHintApplies: true,
+		Type: NewSystemStringType("time_zone"),
+		Default: "SYSTEM",
+	},
+	//TODO: this needs to be a function
+	"timestamp": {
```
Comment (zachmu) [inline sql/system_variables.go:2305]: We have a timestamp function already

File: sql/system_variables.go [hunk 2]
```diff
@@ -0,0 +1,2459 @@
...
+		Name: "timestamp",
+		Scope: SystemVariableScope_Session,
+		Dynamic: true,
+		SetVarHintApplies: true,
+		Type: NewSystemDoubleType("timestamp", 1, 2147483647),
+		Default: float64(1000.0),
```
Comment (zachmu) [inline sql/system_variables.go:2311]: Is this what you mean by "needs to be a function"? Might need to special case it, or make a Default method on the new type interface

File: sql/system_variables.go [hunk 3]
```diff
@@ -0,0 +1,2459 @@
...
+)
+
+// SystemVariable represents a system variable.
+type SystemVariable struct {
+	// Name is the name of the system variable.
+	Name string
+	// Scope defines the scope of the system variable, which is either Global, Session, or Both.
+	Scope SystemVariableScope
+	// Dynamic defines whether the variable may be set during runtime.
+	Dynamic bool
```
Comment (zachmu) [inline sql/system_variables.go:37]: Prefer Writable

File: sql/system_booltype.go [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+	}
+	if expectedVal == 0 {
+		return "0", nil
+	}
+	return "1", nil
+}
+
+// DecodeValue implements SystemVariableType interface.
+func (t systemBoolType) DecodeValue(val string) (interface{}, error) {
+	if val == "0" {
```
Comment (zachmu) [inline sql/system_booltype.go:159]: Many of these accept ON / OFF as well, not sure if you plan to handle this in a higher layer

**Review by zachmu — APPROVED** [review]
This is fine for what it is, but I don't know what you have in mind for the new types, esp. things like encode / decode. Listing the full set of supported system vars is definitely important, so it's good to get that in.

There are a bunch of other outstanding issues with variable assignments, like that we don't differentiate between `@@timestamp` and `@timestamp` for example, either in the analysis logic or the session storage. We really need to do that, since it's totally fine for a single var to have 3 different values: a global system value, a session system value, and a local var value.

Anyway, LGTM, looking forward to the next part

File: sql/system_variables.go (hunk 1 above)
Comment (Hydrocharged) [reply sql/system_variables.go:2305]: I'll fix the comment on this, I don't mean to change this to a function, but rather that `SELECT @@timestamp` should internally call a function as the value isn't static.

File: sql/system_variables.go (hunk 2 above)
Comment (Hydrocharged) [reply sql/system_variables.go:2311]: I think timestamp is the only variable with function-like behavior. But a Default method wouldn't work, as I'm not going to have it as "run default expression unless assigned to at some point", but rather just store the value in a big map and stick the default value in that map. Makes changes and whatnot easier.

File: sql/system_variables.go (hunk 3 above)
Comment (Hydrocharged) [reply sql/system_variables.go:37]: So I'm actually mimicking the terminology used here:
https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html
I can rework the comment, but I'd prefer to keep the terminology the same. Dynamic is used in a lot of places in the documentation.
https://dev.mysql.com/doc/refman/8.0/en/dynamic-system-variables.html

File: sql/system_booltype.go (hunk 4 above)
Comment (Hydrocharged) [reply sql/system_booltype.go:159]: They're handled in `Convert` (which also handles case-insensitive `"true"/"false"`). `DecodeValue/EncodeValue` are just for integrators that want to save the values: it's easier if they only need to worry about a string type. As a result it's adhering to different rules.

**Pull request merged**